### ISSUE TITLE
Nested objective resets are broken

### ISSUE THREAD

**Issue by cdiener** [opencobra/cobrapy#393]
Reproduce:

```Python
In [14]: mod = create_test_model("textbook")

In [15]: mod.objective.expression
Out[15]: -1.0*Biomass_Ecoli_core_reverse_2cdba + 1.0*Biomass_Ecoli_core

In [16]: with mod:
    ...:     with mod: 
    ...:         mod.objective = mod.reactions.ATPM
    ...:     

In [17]: mod.objective.expression
Out[17]: 0
```

**Comment by Midnighter** [comment]
What's `m` in your example?

**Comment by cdiener** [comment]
Ah sorry changed it :)

**Comment by pstjohn** [comment]
I also see it. I'm sure its the call to `set_objective` in https://github.com/opencobra/cobrapy/blob/devel-2/cobra/util/solver.py#L122-L123

I can put in a PR to fix it

**Comment by cdiener** [comment]
Yeah, I think `set_objective` should not call itself recursively. Also if you do could you put the dict on top of the if clause to improve performance a bit?

**Comment by pstjohn** [comment]
So check out the PR: I do think calling set_objective recursively is the best bet here, only because we'd otherwise have to repeat a lot of the if/else code...

**Comment by cdiener** [comment]
Already commented on the PR. Actually said the same thing there :) Also do we really need that additive behavior? It's always much slower than setting the linear coefficients directly and is currently not used anywhere in the code base (always set to false). I think the only thing that is required is to reset all the entire objective to zero before.

**Comment by pstjohn** [comment]
Good point, probably better directed at @hredestig...
I suppose the extra features don't hurt, and speed might be addressed by switching to the `symengine` backend?

**Comment by hredestig** [comment]
We could skip additive behaviour for non-dict values but removing it also for dicts is breaks the 
`
reaction1.objective_coefficient = 1
reaction2.objective_coefficient = 2
` 
where then both reactions ends up in the objective. Perhaps not a big problem and perhaps no one uses this but if we remove `additive` we'd better remove the `reaction.objective_coefficient` setter altogether. I just thought it was possibly a convenient feature and symmetrical, capitalising on the value of expressions. Indeed `symengine` seems likely to render the  speed question moot.

**Comment by hredestig** [comment]
fixed with #395
